tools/build_langserver/langserver/diagnostics.go

```diff
@@ -0,0 +1,341 @@
+package langserver
+
+import (
+	"context"
+	"core"
+	"fmt"
+	"github.com/Workiva/go-datastructures/queue"
+	"parse/asp"
+	"tools/build_langserver/lsp"
+
+	"github.com/sourcegraph/jsonrpc2"
+)
+
+func (h *LsHandler) publishDiagnostics(conn *jsonrpc2.Conn, content string, uri lsp.DocumentURI) error {
+	ctx := context.Background()
+
+	if _, ok := h.workspace.documents[uri]; !ok {
+		return nil
+	}
+
+	h.diagPublisher.diagnose(ctx, h.analyzer, content, uri)
+
+	params := &lsp.PublishDiagnosticsParams{
+		URI:         uri,
+		Diagnostics: h.diagPublisher.stored[uri].stored,
+	}
+	log.Info("Diagnostics detected: %s", params.Diagnostics)
+
+	if err := conn.Notify(ctx, "textDocument/publishDiagnostics", params); err != nil {
+		return err
+	}
+	return nil
+}
+
+type diagnosticsPublisher struct {
+	queue *queue.PriorityQueue
+
+	stored map[lsp.DocumentURI]*diagnosticStore
```

Comment:
why do you need this?

bit concerned that you might need locks around it...

Reply:
Yeah, that's what I was thinking, I might not even need this anymore